File: python/src/kanbus/agent_metadata.py
"""Agent provenance metadata resolution and validation."""

from __future__ import annotations

import json
import os
import re
from typing import Any, Dict, Optional

import click
from pydantic import BaseModel, ConfigDict

from kanbus.ids import format_issue_key
from kanbus.models import AgentMetadata

PLATFORM_PATTERN = re.compile(r"^[a-z0-9_-]{1,64}$")
SECRET_KEY_PATTERN = re.compile(r"(?i)(api[_-]?key|token|secret|password|credential)")
MAX_AGENT_METADATA_BYTES = 2048
MAX_AGENT_NAME_LENGTH = 128

SUGGESTED_AGENT_SETTINGS_KEYS = (
    "temperature",
    "thinking_level",
    "max_output_tokens",
    "speed",
    "reasoning_effort",
)


class AgentMetadataResolutionError(ValueError):
    """Raised when agent metadata cannot be resolved or validated."""


class AgentMetadataRequest(BaseModel):
    """Optional agent metadata inputs from CLI flags and environment."""

    model_config = ConfigDict(extra="forbid")

    platform: Optional[str] = None
    model: Optional[str] = None
    name: Optional[str] = None
    settings_json: Optional[str] = None


def _normalize_optional_text(value: Optional[str]) -> Optional[str]:
    if value is None:
        return None
    stripped = value.strip()
    return stripped if stripped else None


def _normalize_platform(platform: str) -> str:
    slug = re.sub(r"\s+", "_", platform.strip())
    normalized = slug.lower()
    if not PLATFORM_PATTERN.fullmatch(normalized):
        raise AgentMetadataResolutionError("invalid agent platform")
    return normalized


def _normalize_agent_name(name: Optional[str]) -> Optional[str]:
    normalized = _normalize_optional_text(name)
    if normalized is None:
        return None
    if len(normalized) > MAX_AGENT_NAME_LENGTH:
        raise AgentMetadataResolutionError("invalid agent name")
    return normalized


def _validate_settings_keys(settings: Dict[str, Any]) -> None:
    for key in settings:
        if SECRET_KEY_PATTERN.search(key):
            raise AgentMetadataResolutionError(
                "agent settings must not contain secret-like keys"
            )


def _parse_settings_json(settings_json: Optional[str]) -> Dict[str, Any]:
    normalized = _normalize_optional_text(settings_json)
    if normalized is None:
        return {}
    try:
        parsed = json.loads(normalized)
    except json.JSONDecodeError as error:
        raise AgentMetadataResolutionError(
            f"invalid agent settings JSON: {error.msg}"
        ) from error
    if not isinstance(parsed, dict):
        raise AgentMetadataResolutionError(
            "invalid agent settings JSON: expected object"
        )
    return parsed


def _serialize_agent_metadata(agent: AgentMetadata) -> Dict[str, Any]:
    payload = agent.model_dump(exclude_none=True)
    settings = payload.get("settings")
    if isinstance(settings, dict) and not settings:
        payload.pop("settings", None)
    return payload


def build_agent_metadata(
    platform: Optional[str],
    model: Optional[str],
    settings: Optional[Dict[str, Any]] = None,
    name: Optional[str] = None,
) -> Optional[AgentMetadata]:
    """Build validated agent metadata or return None when all inputs are absent.

    :param platform: Agent platform identifier.
    :type platform: Optional[str]
    :param model: Model identifier.
    :type model: Optional[str]
    :param settings: Parsed settings object.
    :type settings: Optional[Dict[str, Any]]
    :param name: Optional session or bot name.
    :type name: Optional[str]
    :return: Validated metadata or None.
    :rtype: Optional[AgentMetadata]
    :raises AgentMetadataResolutionError: If partial or invalid metadata is provided.
    """
    normalized_platform = _normalize_optional_text(platform)
    normalized_model = _normalize_optional_text(model)
    normalized_name = _normalize_agent_name(name)
    settings_payload = settings if settings is not None else {}
    if (
        normalized_platform is None
        and normalized_model is None
        and not settings_payload
        and normalized_name is None
    ):
        return None
    if normalized_platform is None or normalized_model is None:
        raise AgentMetadataResolutionError(
            "agent metadata requires both platform and model"
        )
    _validate_settings_keys(settings_payload)
    agent = AgentMetadata(
        platform=_normalize_platform(normalized_platform),
        model=normalized_model,
        name=normalized_name,
        settings=settings_payload,
    )
    serialized = json.dumps(
        _serialize_agent_metadata(agent),
        separators=(",", ":"),
        sort_keys=True,
    )
    if len(serialized.encode("utf-8")) > MAX_AGENT_METADATA_BYTES:
        raise AgentMetadataResolutionError("agent metadata exceeds maximum size")
    return agent


def resolve_agent_metadata(request: AgentMetadataRequest) -> Optional[AgentMetadata]:
    """Resolve agent metadata from CLI flags with environment defaults.

    :param request: CLI-provided agent metadata inputs.
    :type request: AgentMetadataRequest
    :return: Validated metadata or None.
    :rtype: Optional[AgentMetadata]
    :raises AgentMetadataResolutionError: If validation fails.
    """
    platform = _normalize_optional_text(request.platform)
    model = _normalize_optional_text(request.model)
    name = _normalize_agent_name(request.name)
    settings_json = _normalize_optional_text(request.settings_json)
    if platform is None:
        platform = _normalize_optional_text(os.getenv("KANBUS_AGENT_PLATFORM"))
    if model is None:
        model = _normalize_optional_text(os.getenv("KANBUS_AGENT_MODEL"))
    if name is None:
        name = _normalize_agent_name(os.getenv("KANBUS_AGENT_NAME"))
    if settings_json is None:
        settings_json = _normalize_optional_text(os.getenv("KANBUS_AGENT_SETTINGS"))
    settings = _parse_settings_json(settings_json)
    return build_agent_metadata(platform, model, settings, name)


def format_agent_display_line(agent: AgentMetadata) -> str:
    """Format agent metadata for compact CLI display.

    :param agent: Agent metadata to format.
    :type agent: AgentMetadata
    :return: Display string such as ``cloud-agent / cursor / composer-2.5``.
    :rtype: str
    """
    platform_model = f"{agent.platform} / {agent.model}"
    if agent.name:
        return f"{agent.name} / {platform_model}"
    return platform_model


def _format_setting_value(value: Any) -> str:
    if isinstance(value, (dict, list)):
        return json.dumps(value, separators=(",", ":"), sort_keys=True)
    if isinstance(value, bool):
        return "true" if value else "false"
    if value is None:
        return "null"
    return str(value)


def format_agent_settings_display(agent: AgentMetadata) -> Optional[str]:
    """Format agent settings for CLI display.

    :param agent: Agent metadata to format.
    :type agent: AgentMetadata
    :return: Comma-separated key=value pairs or None when empty.
    :rtype: Optional[str]
    """
    if not agent.settings:
        return None
    parts = [
        f"{key}={_format_setting_value(value)}"
        for key, value in sorted(agent.settings.items())
    ]
    return ", ".join(parts)


def agent_metadata_to_event_value(agent: AgentMetadata) -> Dict[str, Any]:
    """Serialize agent metadata for event payloads.

    :param agent: Agent metadata to serialize.
    :type agent: AgentMetadata
    :return: JSON-compatible payload fragment.
    :rtype: Dict[str, Any]
    """
    return _serialize_agent_metadata(agent)


EXAMPLE_AGENT_PLATFORM = "Cursor"
EXAMPLE_AGENT_MODEL = "Composer 2.5"
EXAMPLE_AGENT_NAME = "Cloud Agent"
AGENT_METADATA_ALREADY_SET = "agent metadata is already set"


def agent_provenance_is_complete(agent: Optional[AgentMetadata]) -> bool:
    """Return whether platform, model, and session name are all present.

    :param agent: Stored agent metadata.
    :type agent: Optional[AgentMetadata]
    :return: True when tagging is complete.
    :rtype: bool
    """
    return agent is not None and bool(agent.name)


def missing_agent_provenance_fields(agent: Optional[AgentMetadata]) -> list[str]:
    """List required provenance fields that are not set.

    :param agent: Stored or resolved agent metadata.
    :type agent: Optional[AgentMetadata]
    :return: Missing field names in stable order.
    :rtype: list[str]
    """
    if agent is None:
        return ["platform", "model", "name"]
    if agent.name:
        return []
    return ["name"]


def assign_agent_metadata_if_incomplete(
    existing: Optional[AgentMetadata], incoming: Optional[AgentMetadata]
) -> Optional[AgentMetadata]:
    """Set agent metadata when the existing record is incomplete.

    :param existing: Current issue or comment agent metadata.
    :type existing: Optional[AgentMetadata]
    :param incoming: Newly resolved metadata, or None when not requested.
    :type incoming: Optional[AgentMetadata]
    :return: Metadata to store.
    :rtype: Optional[AgentMetadata]
    :raises AgentMetadataResolutionError: If complete metadata would be replaced.
    """
    if incoming is None:
        return existing
    if agent_provenance_is_complete(existing):
        raise AgentMetadataResolutionError(AGENT_METADATA_ALREADY_SET)
    return incoming


def format_agent_provenance_warning(
    issue_identifier: str,
    agent: Optional[AgentMetadata],
    comment_id: Optional[str] = None,
) -> str:
    """Build a warning that includes a one-step follow-up command.

    :param issue_identifier: Issue identifier to include in the follow-up.
    :type issue_identifier: str
    :param agent: Metadata already stored or resolved for this write.
    :type agent: Optional[AgentMetadata]
    :param comment_id: Comment identifier when warning about a comment.
    :type comment_id: Optional[str]
    :return: Multi-line warning text.
    :rtype: str
    """
    missing = missing_agent_provenance_fields(agent)
    missing_text = ", ".join(missing)
    issue_key = format_issue_key(issue_identifier, False)
    if comment_id:
        command = f"kbs comment update {issue_key} {comment_id}"
    else:
        command = f"kbs update {issue_key}"
    if "platform" in missing:
        command = f'{command} --agent-platform "{EXAMPLE_AGENT_PLATFORM}"'
    if "model" in missing:
        command = f'{command} --agent-model "{EXAMPLE_AGENT_MODEL}"'
    if "name" in missing:
        command = f'{command} --agent-name "{EXAMPLE_AGENT_NAME}"'
    return (
        f"WARNING: agent provenance is incomplete (missing: {missing_text}).\n"
        "See CONTRIBUTING_AGENT.md (Agent provenance metadata).\n"
        "To add it in one step:\n"
        f"  {command}\n"
        "To skip tagging this time, re-run with --no-agent-provenance."
    )


def emit_agent_provenance_warning(
    issue_identifier: str,
    agent: Optional[AgentMetadata],
    comment_id: Optional[str] = None,
    silenced: bool = False,
) -> None:
    """Write the incomplete-provenance warning to stderr when needed.

    :param issue_identifier: Issue identifier to include in the follow-up.
    :type issue_identifier: str
    :param agent: Metadata already stored or resolved for this write.
    :type agent: Optional[AgentMetadata]
    :param comment_id: Comment identifier when warning about a comment.
    :type comment_id: Optional[str]
    :param silenced: When True, do not emit a warning.
    :type silenced: bool
    """
    if silenced or agent_provenance_is_complete(agent):
        return
    click.echo(
        format_agent_provenance_warning(issue_identifier, agent, comment_id),
        err=True,
    )


def reject_agent_metadata_in_beads_mode(agent_present: bool) -> None:
    """Reject agent metadata mutations in Beads compatibility mode.

    :param agent_present: Whether agent metadata was supplied.
    :type agent_present: bool
    :raises AgentMetadataResolutionError: When agent metadata is supplied in Beads mode.
    """
    if agent_present:
        raise AgentMetadataResolutionError(
            "agent metadata requires native Kanbus issue storage"
        )

File: rust/src/agent_metadata.rs
//! Agent provenance metadata resolution and validation.

use regex::Regex;
use serde_json::Value;
use std::collections::BTreeMap;
use std::sync::LazyLock;

use crate::error::KanbusError;
use crate::models::AgentMetadata;

static PLATFORM_PATTERN: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^[a-z0-9_-]{1,64}$").expect("platform regex"));
static SECRET_KEY_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(?i)(api[_-]?key|token|secret|password|credential)").expect("secret key regex")
});

const MAX_AGENT_METADATA_BYTES: usize = 2048;
const MAX_AGENT_NAME_LENGTH: usize = 128;

/// Optional agent metadata inputs from CLI flags and environment.
#[derive(Debug, Clone, Default)]
pub struct AgentMetadataRequest {
    pub platform: Option<String>,
    pub model: Option<String>,
    pub name: Option<String>,
    pub settings_json: Option<String>,
}

fn normalize_optional_text(value: Option<&str>) -> Option<String> {
    value.and_then(|text| {
        let trimmed = text.trim();
        if trimmed.is_empty() {
            None
        } else {
            Some(trimmed.to_string())
        }
    })
}

fn normalize_platform(platform: &str) -> Result<String, KanbusError> {
    let slug = platform.split_whitespace().collect::<Vec<_>>().join("_");
    let normalized = slug.to_ascii_lowercase();
    if !PLATFORM_PATTERN.is_match(&normalized) {
        return Err(KanbusError::IssueOperation(
            "invalid agent platform".to_string(),
        ));
    }
    Ok(normalized)
}

fn normalize_agent_name(name: Option<&str>) -> Result<Option<String>, KanbusError> {
    let Some(normalized) = normalize_optional_text(name) else {
        return Ok(None);
    };
    if normalized.len() > MAX_AGENT_NAME_LENGTH {
        return Err(KanbusError::IssueOperation(
            "invalid agent name".to_string(),
        ));
    }
    Ok(Some(normalized))
}

fn validate_settings_keys(settings: &BTreeMap<String, Value>) -> Result<(), KanbusError> {
    for key in settings.keys() {
        if SECRET_KEY_PATTERN.is_match(key) {
            return Err(KanbusError::IssueOperation(
                "agent settings must not contain secret-like keys".to_string(),
            ));
        }
    }
    Ok(())
}

fn parse_settings_json(
    settings_json: Option<&str>,
) -> Result<BTreeMap<String, Value>, KanbusError> {
    let Some(normalized) = normalize_optional_text(settings_json) else {
        return Ok(BTreeMap::new());
    };
    let parsed: Value = serde_json::from_str(&normalized).map_err(|error| {
        KanbusError::IssueOperation(format!("invalid agent settings JSON: {error}"))
    })?;
    let Value::Object(object) = parsed else {
        return Err(KanbusError::IssueOperation(
            "invalid agent settings JSON: expected object".to_string(),
        ));
    };
    Ok(object.into_iter().collect())
}

/// Build validated agent metadata or return None when all inputs are absent.
pub fn build_agent_metadata(
    platform: Option<&str>,
    model: Option<&str>,
    settings: &BTreeMap<String, Value>,
    name: Option<&str>,
) -> Result<Option<AgentMetadata>, KanbusError> {
    let normalized_platform = normalize_optional_text(platform);
    let normalized_model = normalize_optional_text(model);
    let normalized_name = normalize_agent_name(name)?;
    if normalized_platform.is_none()
        && normalized_model.is_none()
        && settings.is_empty()
        && normalized_name.is_none()
    {
        return Ok(None);
    }
    let platform_value = normalized_platform.ok_or_else(|| {
        KanbusError::IssueOperation("agent metadata requires both platform and model".to_string())
    })?;
    let model_value = normalized_model.ok_or_else(|| {
        KanbusError::IssueOperation("agent metadata requires both platform and model".to_string())
    })?;
    validate_settings_keys(settings)?;
    let agent = AgentMetadata {
        platform: normalize_platform(&platform_value)?,
        model: model_value,
        name: normalized_name,
        settings: settings.clone(),
    };
    let serialized = serde_json::to_string(&agent)
        .map_err(|error| KanbusError::IssueOperation(format!("invalid agent metadata: {error}")))?;
    if serialized.len() > MAX_AGENT_METADATA_BYTES {
        return Err(KanbusError::IssueOperation(
            "agent metadata exceeds maximum size".to_string(),
        ));
    }
    Ok(Some(agent))
}

/// Resolve agent metadata from CLI flags with environment defaults.
pub fn resolve_agent_metadata(
    request: &AgentMetadataRequest,
) -> Result<Option<AgentMetadata>, KanbusError> {
    let mut platform = normalize_optional_text(request.platform.as_deref());
    let mut model = normalize_optional_text(request.model.as_deref());
    let mut name = normalize_agent_name(request.name.as_deref())?;
    let mut settings_json = normalize_optional_text(request.settings_json.as_deref());
    if platform.is_none() {
        platform = normalize_optional_text(std::env::var("KANBUS_AGENT_PLATFORM").ok().as_deref());
    }
    if model.is_none() {
        model = normalize_optional_text(std::env::var("KANBUS_AGENT_MODEL").ok().as_deref());
    }
    if name.is_none() {
        name = normalize_agent_name(std::env::var("KANBUS_AGENT_NAME").ok().as_deref())?;
    }
    if settings_json.is_none() {
        settings_json =
            normalize_optional_text(std::env::var("KANBUS_AGENT_SETTINGS").ok().as_deref());
    }
    let settings = parse_settings_json(settings_json.as_deref())?;
    build_agent_metadata(
        platform.as_deref(),
        model.as_deref(),
        &settings,
        name.as_deref(),
    )
}

fn format_setting_value(value: &Value) -> String {
    match value {
        Value::String(text) => text.clone(),
        Value::Bool(flag) => flag.to_string(),
        Value::Null => "null".to_string(),
        other => other.to_string(),
    }
}

/// Format agent metadata for compact CLI display.
pub fn format_agent_display_line(agent: &AgentMetadata) -> String {
    let platform_model = format!("{} / {}", agent.platform, agent.model);
    if let Some(name) = agent.name.as_deref() {
        format!("{name} / {platform_model}")
    } else {
        platform_model
    }
}

/// Format agent settings for CLI display.
pub fn format_agent_settings_display(agent: &AgentMetadata) -> Option<String> {
    if agent.settings.is_empty() {
        return None;
    }
    let parts = agent
        .settings
        .iter()
        .map(|(key, value)| format!("{key}={}", format_setting_value(value)))
        .collect::<Vec<_>>();
    Some(parts.join(", "))
}

/// Serialize agent metadata for event payloads.
pub fn agent_metadata_to_event_value(agent: &AgentMetadata) -> Value {
    serde_json::to_value(agent).unwrap_or(Value::Null)
}

/// Example product name used in follow-up commands.
pub const EXAMPLE_AGENT_PLATFORM: &str = "Cursor";
/// Example model name used in follow-up commands.
pub const EXAMPLE_AGENT_MODEL: &str = "Composer 2.5";
/// Example session name used in follow-up commands.
pub const EXAMPLE_AGENT_NAME: &str = "Cloud Agent";
/// Error when complete agent metadata would be replaced.
pub const AGENT_METADATA_ALREADY_SET: &str = "agent metadata is already set";

/// Return whether platform, model, and session name are all present.
pub fn agent_provenance_is_complete(agent: Option<&AgentMetadata>) -> bool {
    agent.is_some_and(|metadata| {
        metadata
            .name
            .as_deref()
            .is_some_and(|name| !name.is_empty())
    })
}

/// List required provenance fields that are not set.
pub fn missing_agent_provenance_fields(agent: Option<&AgentMetadata>) -> Vec<&'static str> {
    match agent {
        None => vec!["platform", "model", "name"],
        Some(metadata)
            if metadata
                .name
                .as_deref()
                .is_some_and(|name| !name.is_empty()) =>
        {
            Vec::new()
        }
        Some(_) => vec!["name"],
    }
}

/// Set agent metadata when the existing record is incomplete.
///
/// # Errors
/// Returns `KanbusError` when complete metadata would be replaced.
pub fn assign_agent_metadata_if_incomplete(
    existing: Option<AgentMetadata>,
    incoming: Option<AgentMetadata>,
) -> Result<Option<AgentMetadata>, KanbusError> {
    let Some(incoming_metadata) = incoming else {
        return Ok(existing);
    };
    if agent_provenance_is_complete(existing.as_ref()) {
        return Err(KanbusError::IssueOperation(
            AGENT_METADATA_ALREADY_SET.to_string(),
        ));
    }
    Ok(Some(incoming_metadata))
}

/// Build a warning that includes a one-step follow-up command.
pub fn format_agent_provenance_warning(
    issue_identifier: &str,
    agent: Option<&AgentMetadata>,
    comment_id: Option<&str>,
) -> String {
    let missing = missing_agent_provenance_fields(agent);
    let missing_text = missing.join(", ");
    let issue_key = crate::ids::format_issue_key(issue_identifier, false);
    let mut command = if let Some(comment_identifier) = comment_id {
        format!("kbs comment update {issue_key} {comment_identifier}")
    } else {
        format!("kbs update {issue_key}")
    };
    if missing.contains(&"platform") {
        command.push_str(&format!(" --agent-platform \"{EXAMPLE_AGENT_PLATFORM}\""));
    }
    if missing.contains(&"model") {
        command.push_str(&format!(" --agent-model \"{EXAMPLE_AGENT_MODEL}\""));
    }
    if missing.contains(&"name") {
        command.push_str(&format!(" --agent-name \"{EXAMPLE_AGENT_NAME}\""));
    }
    format!(
        "WARNING: agent provenance is incomplete (missing: {missing_text}).\n\
See CONTRIBUTING_AGENT.md (Agent provenance metadata).\n\
To add it in one step:\n\
  {command}\n\
To skip tagging this time, re-run with --no-agent-provenance."
    )
}

/// Write the incomplete-provenance warning to stderr when needed.
pub fn emit_agent_provenance_warning(
    issue_identifier: &str,
    agent: Option<&AgentMetadata>,
    comment_id: Option<&str>,
    silenced: bool,
) {
    if silenced || agent_provenance_is_complete(agent) {
        return;
    }
    crate::rich_text_signals::emit_stderr_message(&format_agent_provenance_warning(
        issue_identifier,
        agent,
        comment_id,
    ));
}

/// Reject agent metadata mutations in Beads compatibility mode.
pub fn reject_agent_metadata_in_beads_mode(agent_present: bool) -> Result<(), KanbusError> {
    if agent_present {
        return Err(KanbusError::IssueOperation(
            "agent metadata requires native Kanbus issue storage".to_string(),
        ));
    }
    Ok(())
}

/// Format comment author label with optional agent suffix.
pub fn format_comment_author_label(author: &str, agent: Option<&AgentMetadata>) -> String {
    match agent {
        Some(metadata) => format!("{author} ({}):", format_agent_display_line(metadata)),
        None => format!("{author}:"),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn build_agent_metadata_requires_platform_and_model_together() {
        let error = build_agent_metadata(Some("cursor"), None, &BTreeMap::new(), None)
            .expect_err("expected error");
        assert_eq!(
            error.to_string(),
            "agent metadata requires both platform and model"
        );
    }

    #[test]
    fn reject_secret_like_settings_keys() {
        let mut settings = BTreeMap::new();
        settings.insert(
            "openai_api_key".to_string(),
            Value::String("secret".to_string()),
        );
        let error = build_agent_metadata(Some("cursor"), Some("model"), &settings, None)
            .expect_err("expected error");
        assert_eq!(
            error.to_string(),
            "agent settings must not contain secret-like keys"
        );
    }

    #[test]
    fn title_case_platform_with_spaces_normalizes() {
        let agent = build_agent_metadata(
            Some("Claude Code"),
            Some("Composer 2.5"),
            &BTreeMap::new(),
            None,
        )
        .expect("expected metadata")
        .expect("metadata present");
        assert_eq!(agent.platform, "claude_code");
        assert_eq!(agent.model, "Composer 2.5");
    }

    #[test]
    fn format_agent_provenance_warning_includes_follow_up() {
        let warning = format_agent_provenance_warning("kanbus-aaa", None, None);
        assert!(warning.contains("agent provenance is incomplete (missing: platform, model, name)"));
        assert!(warning.contains("CONTRIBUTING_AGENT.md"));
        assert!(warning.contains("kbs update kanbus-aaa --agent-platform \"Cursor\""));
        assert!(warning.contains("--agent-name \"Cloud Agent\""));
    }

    #[test]
    fn accept_unknown_settings_keys_and_values() {
        let mut settings = BTreeMap::new();
        settings.insert(
            "reasoning_effort".to_string(),
            Value::String("turbo".to_string()),
        );
        let agent = build_agent_metadata(Some("cursor"), Some("model"), &settings, None)
            .expect("expected metadata");
        assert_eq!(
            agent
                .expect("metadata present")
                .settings
                .get("reasoning_effort")
                .and_then(|value| value.as_str()),
            Some("turbo")
        );
    }
}
